front: enable occurrence selection for path projection

clarani opened the pull request:
closes #11476

Rework of https://github.com/OpenRailAssociation/osrd/pull/12222 (first part merged in https://github.com/OpenRailAssociation/osrd/pull/12882)


## ACs
- [ ] You can project onto an occurrence (see mockup for the icon to indicate that you are projecting onto a mission/occurrence):
  - [ ] If you project onto the model train or onto an occurrence that has the same path or schedule as the model train: blue projection icon on the mission's parent map (folded AND unfolded mode). Nothing on the occurrence
  - [ ] If we project onto an occurrence that does not have the same path or schedule as the model train: gray projection icon on the model train (folded AND unfolded mode) + blue projection icon on the occurrence onto which we are projecting
- [ ] Nothing special happens if you deactivate the occurrence used for the projection (it does not change the space time chart, except the current occurrence is hidden)

## Screenshots
<img width="406" height="450" alt="image" src="https://github.com/user-attachments/assets/9d9256dc-6ca4-472e-a951-6f50431158b9" />

anisometropie committed:
front: enable projection on occurrence

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

anisometropie committed:
front: display projection icon when projecting on occurrences

If the path of an occurrence is used for projection:
- its paced train necessarily has a projection icon, which is blue if the occurrence in question has no path/schedule exception, gray otherwise
- the occurrence has a projection icon only if it is a path/schedule exception

We will deal with cases where the occurrence is only a schedule exception and its path does not differ from that of the paced train later.

Signed-off-by: Valentin Chanas <anisometropie@gmail.com>

clarani committed:
front: trainIdUsedForProjection can be an OccurrenceId

Until now, trainIdUsedForProjection could only be a trainScheduleId or a pacedTrainId,
which limited us from using an exception (occurrence) to project the trains.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

RomainValls reviewed (approved):
Tested, LGTM 👍

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -247,3 +247,12 @@ export const extractExceptionIdFromOccurrenceId = (occurrenceId: OccurrenceId):
 };
+
+/**
+ * Given a paced train id with a PacedTrainId format (used across the front),
```
An occurrence ID is given, not a paced train ID.

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -247,3 +247,12 @@ export const extractExceptionIdFromOccurrenceId = (occurrenceId: OccurrenceId):
+ * Given a paced train id with a PacedTrainId format (used across the front),
+ * returns the train id in the Editoast format (used for api).
+ */
+export const extractEditoastIdFromOccurrenceId = (occurrenceId: OccurrenceId): number => {
```
I find the name of this helper function ambiguous: this makes it sound like it returns an editoast ID representing the occurrence, but it returns an editoast ID representing the paced train.

Maybe we should take inspiration from the existing `extractPacedTrainIdFromOccurrenceId()` function, and call this one e.g. `extractEditoastPacedTrainIdFromOccurrenceId`?

Alternatively, I think this helper isn't really saving that much: calling the two functions one after the other would be perfectly fine.

clarani replied:
Done

emersion commented on front/src/modules/timetableItem/components/Timetable/types.ts:
```diff
@@ -119,6 +119,7 @@ export type Occurrence = {
   rollingStock?: LightRollingStockWithLiveries;
   startTime: Date;
   stopsCount: number;
+  exceptionKey?: string;
```
I think adding an `exceptionKey` field here in `Occurrence` is a good idea, but it doesn't seem like this new field is used in this commit?

clarani replied:
Hum it is used, when selecting the occurrence to be used for projection (in `selectOccurrenceForProjection`, in `useOccurrenceActions`).

emersion replied:
This comment is about the `Occurrence` type. `updateTrainIdUsedForProjection()` updates the store's `trainUsedForProjection` field. These two are separate.

clarani replied:
Hum it was used for the selection of the projection like this:
```
const selectOccurrenceForProjection = useCallback((occurrence: Occurrence) => {
  dispatch(updateTrainIdUsedForProjection({ id: occurrence.id, exceptionKey: occurrence.key }));
}, []);
```

But anyway, I droped the exceptionKey in updateTrainIdUsedForProjection, so I can drop this field too now :+1:

emersion commented on front/src/modules/timetableItem/components/Timetable/PacedTrain/hooks/useOccurrenceActions.ts:
```diff
@@ -140,13 +140,13 @@ const useOccurrenceActions = ({
-        // TODO exceptions : update projected occurrence id in issue https://github.com/OpenRailAssociation/osrd/issues/11476
+        if (selectedTrainId === occurrence.id) {
+          dispatch(updateSelectedTrainId(firstEnabledOccurrence?.id));
+        }
```
I don't understand this change. The new code seems completely equivalent to the old one, exception that we do a `occurrences.find()` call unnecessarily?

clarani replied:
Hum you're right, I revert this change

emersion commented on front/src/reducers/simulationResults/types.ts:
```diff
@@ -33,6 +33,9 @@ export type ProjectionType = 'trackProjection' | 'operationalPointProjection';
-  trainIdUsedForProjection?: TrainScheduleId | PacedTrainId | OccurrenceId;
+  trainUsedForProjection?: {
+    id: TrainScheduleId | PacedTrainId | OccurrenceId;
+    exceptionKey?: string;
```
I'm not a fan of storing the `exceptionKey` in the `trainUsedForProjection`:

- We need to add special cases when the exception is removed, to clear the `exceptionKey` but keep any potential occurrence index in  `OccurrenceId`.
- This potentially duplicates information and allows for inconsistent state, e.g. `exception_42_foo` in `id` (indicating the "foo" exception key) and `bar` in `exceptionKey`.

Could we instead grab the exception ID from the `Occurrence`?

clarani replied:
I'm not a fan either, but I need the key to then call the path endpoint and I can't get the exception key easily for indexedOccurrences...

The alternative would be to store only the trainId, but that adds processing for indexedOccurrences, since it would be necessary to:
- retrieve the complete pacedTrain with exceptions (from the API or from timetableItems ?)
- check the exceptions to see if this occurrence is one or not
- retrieve the key if it is an exception, otherwise use the pacedTrain

I thought it would be easier to manage the key directly in the store, but we can change that if you want

emersion replied:
It shouldn't be necessary to have that much logic: we only need to look up the exception if the ID is a `IndexedOccurrenceId`. Also, we shouldn't need to fetch anything from the API - `timetableItems` from `useScenarioData` should be enough.

The following could be put into a re-usable helper if need be:

```ts
  const occurrenceId: IndexedOccurrenceId;
  const pacedTrain = timetableItemsById.get(extractPacedTrainIdFromOccurrenceId(occurrenceId));
  if (!pacedTrain || !isPacedTrainResponseWithPacedTrainId(pacedTrain)) {
    throw new Error('oops');
  }
  const occurrenceIndex = extractOccurrenceIndexFromOccurrenceId(occurrenceId);
  const exception = pacedTrain.exceptions.find((exception) => exception.occurrence_index === occurrenceIndex);
  const key = exception?.key;
```

clarani replied:
You were right, it was not as difficult as I thought, it's way cleaner like this. Thanks ! :pray:

emersion commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -155,13 +156,15 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+    if (!trainIdUsedForProjection) return undefined;
+
+    const timetableItemId = isOccurrenceId(trainIdUsedForProjection)
+      ? extractPacedTrainIdFromOccurrenceId(trainIdUsedForProjection)
```
Hm, here we grab the base paced train and ignore exceptions. This means any changes made by the exception are not reflected in `timetableItemUsedForProjection`. What if an exception has a different path, for instance?

Among other things, in `SimulationResults` we call `useGetProjectedTrainOperationalPoints()` with `projectionData?.timetableItem`. Wouldn't this result in the wrong OP list being computed?

clarani replied:
Hum yes you're right, this is very confusing...

clarani replied:
Done

emersion reviewed (commented):
Here are a few comments!

clarani committed:
front: rename props path->pathfinding in ProjectionData

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: use ProjectionData pathfinding in useGetProjectedTrainOperationalPoints

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: store the path used for projection in ProjectionData

Previously, we stored the timetable item in ProjectionData, but we only really used its path in the rest of the code.
We might as well only store the path, which will make it easier for us to make changes in future commits.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: remove useless condition in useGetProjectedTrainOperationalPoints

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

Math-R commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -154,10 +160,26 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+
+    const pacedTrainId = extractPacedTrainIdFromOccurrenceId(trainIdUsedForProjection);
+    const pacedTrain = timetableItemsById.get(pacedTrainId);
+    if (!pacedTrain || !isPacedTrainResponseWithPacedTrainId(pacedTrain)) throw new Error();
```
Maybe you could contextualize the error with a message to be more clearer

clarani replied:
Done :white_check_mark:

Math-R commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -154,10 +160,26 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+    }
+
+    const pacedTrainId = extractPacedTrainIdFromOccurrenceId(trainIdUsedForProjection);
+    const pacedTrain = timetableItemsById.get(pacedTrainId);
```
it hink you can group the two `get` by using a ternary

```
const train = timetableItemsById.get(
    isOccurrenceId(trainIdUsedForProjection)
      ? extractPacedTrainIdFromOccurrenceId(trainIdUsedForProjection)
      : trainIdUsedForProjection
  );
```

clarani replied:
I have modified a bit the logic of this useMemo, tell me if it's fine for you

Math-R commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -154,10 +160,26 @@ const useScenarioData = (scenario: ScenarioResponse, infra: InfraWithState) => {
+    const pacedTrain = timetableItemsById.get(pacedTrainId);
+    if (!pacedTrain || !isPacedTrainResponseWithPacedTrainId(pacedTrain)) throw new Error();
+
+    let exception: PacedTrainException | undefined;
+    if (isIndexedOccurrenceId(trainIdUsedForProjection)) {
+      const index = extractOccurrenceIndexFromOccurrenceId(trainIdUsedForProjection);
+      exception = pacedTrain.exceptions.find((e) => e.occurrence_index === index);
+    } else {
+      const key = extractExceptionIdFromOccurrenceId(trainIdUsedForProjection);
+      exception = pacedTrain.exceptions.find((e) => e.key === key);
+    }
```
I think you could simplify this block by extracting the identifier (index or key) using a ternary expression, and then using a single find() call with a conditional property check.
This would avoid the duplication and make the logic a bit easier to follow.

emersion replied:
I must admit, I personally find the current version clearer, since there is so little common code between these two branches.

clarani replied:
I also find my version clearer, can we keep it like this ?

Math-R replied:
```
const isIndexed = isIndexedOccurrenceId(trainIdUsedForProjection);
const id = isIndexed
  ? extractOccurrenceIndexFromOccurrenceId(trainIdUsedForProjection)
  : extractExceptionIdFromOccurrenceId(trainIdUsedForProjection);

const exception = pacedTrain.exceptions.find((e) =>
  isIndexed ? e.occurrence_index === id : e.key === id
);
```

i was suggesting something like this. But i'm okay if you prefer to keep your version.

clarani replied:
Ok, then I'll keep mine :) Thanks

Math-R reviewed (approved):
lgtm
